### PR TITLE
Features/action excel linux transformation

### PR DESCRIPTION
**Pull request by uzanh**

### CODE REVIEW

File: Ginger/GingerCoreNET/GingerCoreNET.csproj [hunk 1]
```diff
@@ -247,8 +248,10 @@
     <PackageReference Include="Selenium.WebDriver.MicrosoftWebDriver" Version="10.0.17134" />
     <PackageReference Include="SharpAdbClient" Version="2.3.23" />
     <PackageReference Include="SSH.NET" Version="2020.0.1" />
+    <PackageReference Include="System.Data.DataSetExtensions" Version="4.5.0" />
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/GingerCoreNET.csproj:251]: why those System.Data.DataSetExtensions and System.Data.OleDb were added?

File: Ginger/GingerCoreNET/Run/Excels/ExcelNPOIOperations.cs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+                    }
+                    dtExcelTable.Rows.Add(dr);
+                    i++;
+                    currentRow = sheet.GetRow(i);
+                }
+                return dtExcelTable;
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Can't convert sheet to data table, " + ex.Message);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNET/Run/Excels/ExcelNPOIOperations.cs:89]: No need to catch and throw, just let the exception get to the higher caller to be handeled

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                TestResources.GetTestResourcesFile(@"Excel" + Path.DirectorySeparatorChar + "Names.xlsx"));
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue("SheetName","Sheet1");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadData;
+            actExcel.AddNewReturnParams = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(actExcel.ActReturnValues.Count,4);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:44]: Add validation of actual returned values for all unit tests

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue("SheetName", "Sheet1");
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue("SelectRowsWhere", "Last='Cohen'");
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue("PrimaryKeyColumn", "ID");
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue("ColMappingRules", "First='Marco'");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.WriteData;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:176]: Need to validate that the expected values were wrote to the copy/temp excel file by reading it as you wish

**Review by MeniKadosh1 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Ginger.csproj [hunk 5]
```diff
@@ -4765,6 +4765,9 @@
     <PackageReference Include="NSwag.Core.Yaml">
       <Version>12.0.9</Version>
     </PackageReference>
+    <PackageReference Include="Oracle.ManagedDataAccess">
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/Ginger/Ginger.csproj:4768]: ambar add it in our meeting , for his changes in DotNetFrameworkHelper

File: Ginger/GingerCore/GingerCore.csproj [hunk 6]
```diff
@@ -848,9 +847,6 @@
     <PackageReference Include="Npgsql">
       <Version>4.0.7</Version>
     </PackageReference>
-    <PackageReference Include="Oracle.ManagedDataAccess">
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCore/GingerCore.csproj:851]: ambar ask to remove it in our meeting

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+            TBH.AddLineBreak();
+            TBH.AddLineBreak();
+            TBH.AddText("Read Excel :- If you want to read excel sheet from system then select Read data from Excel Action type dropdown, Then browse the file by clicking Browse button.Once you " +
+             "browse the file then all the sheets will get bind to Sheet Name dropdown,Select sheet name and click on view button.If you want to put where condition on excel sheet then type column name and it's value e.g. ColName=1 for integer column and ColName='colname' for string column");
+            TBH.AddLineBreak();
+            TBH.AddLineBreak();
+            TBH.AddText("Write Excel :- Similar as read action,to write an excel, select Write data from Excel Action type dropdown, Then browse the file by clicking Browse button,select sheet name and then add where condition on column which you want to write, Then add primary key column if you have any.Then finally to write an excel add column name =Variable name in Variable to col textbox and run the action");
+            TBH.AddLineBreak();
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs:55]: Done

File: Ginger/GingerConsole/GingerConsole.csproj [hunk 8]
```diff
@@ -33,7 +33,7 @@
     <PackageReference Include="Ginger.ExecuterService.Contracts" Version="1.0.8" />
     <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="3.21.1" />
     <PackageReference Include="System.Resources.Extensions" Version="4.7.1" />
-    <PackageReference Include="System.Security.Permissions" Version="4.7.0" />
+    <PackageReference Include="System.Security.Permissions" Version="5.0.0" />
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerConsole/GingerConsole.csproj:36]: NPOI nuget dependency

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 4 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:176]: done

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 3 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:44]: done

File: Ginger/GingerCoreNET/Run/Excels/ExcelNPOIOperations.cs (hunk 2 above)
Comment (uzanh) [reply Ginger/GingerCoreNET/Run/Excels/ExcelNPOIOperations.cs:89]: done

File: Ginger/GingerCoreNET/GingerCoreNET.csproj (hunk 1 above)
Comment (uzanh) [reply Ginger/GingerCoreNET/GingerCoreNET.csproj:251]: System.Data.OleDb removed
System.Data.DataSetExtensions for datatable

File: Ginger/Ginger/Actions/ActionEditPages/ActExcelEditPage.xaml.cs [hunk 9]
```diff
@@ -112,20 +116,44 @@ private void ViewDataButton_Click(object sender, RoutedEventArgs e)
...
+                Reporter.ToUser(eUserMsgKey.ExcelInvalidFieldData);
+                return;
+            }
+            ExcelDataGrid.ItemsSource = dt.AsDataView();
         }
 
         private void ViewWhereButton_Click(object sender, RoutedEventArgs e)
         {
+            if (!mAct.CheckMandatoryFieldsExists(new List<string>() { 
+                nameof(mAct.CalculatedFileName), nameof(mAct.CalculatedSheetName),  nameof(mAct.CalculatedFilter)}))
```
Comment (MeniKadosh1) [inline Ginger/Ginger/Actions/ActionEditPages/ActExcelEditPage.xaml.cs:131]: you tested getting the calculated fields value on edit page without running the Action first?

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+                    ReadData();
+                    break;
+                case eExcelActionType.WriteData:
+                    WriteData();
+                    break;
+                case eExcelActionType.ReadCellData:
+                    ReadCellData();
+                    break;
+                default:
+                    break;
```
[the remark this answers is not in the document]
Comment (MeniKadosh1) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs:233]: default value with error is still missing

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+                    GetInputParamCalculatedValue(nameof(ExcelFileName)));
+            string sheetName = GetInputParamCalculatedValue(nameof(SheetName));
+            string selectRowsWhere = GetInputParamCalculatedValue(nameof(SelectRowsWhere));
+            DataTable excelDataTable = excelOperator.ReadCellData(excelFileName, sheetName, selectRowsWhere, SelectAllRows);
+            try
+            {
+                if (!string.IsNullOrEmpty(SelectRowsWhere) && !SelectAllRows)
+                {
+                    string CellValue = excelDataTable.Rows[0][0].ToString();
+                    AddOrUpdateReturnParamActual("Actual", CellValue);
```
[the remark this answers is not in the document]
Comment (MeniKadosh1) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs:262]: I still shows "Actual" and not column name

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SheetName), "Sheet1");
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SelectRowsWhere), "Last='Cohen'");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadData;
+            actExcel.AddNewReturnParams = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(string.Join(',',actExcel.ActReturnValues.Select(x=> x.Actual).ToList()), "1,Mark,Cohen,2109 Fox Dr");
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:62]: this is not enough to validate only the "Actual" field, need to validate also the "Param" and "Path"

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SheetName), "Sheet1");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadData;
+            actExcel.AddNewReturnParams = true;
+            actExcel.SelectAllRows = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(string.Join(',', actExcel.ActReturnValues.Select(x => x.Actual).ToList()), "3,Mike,Bond,AZ");
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:82]: this is not enough to validate only the "Actual" field, need to validate also the "Param" and "Path"

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SelectRowsWhere), "Last='Cohen'");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadData;
+            actExcel.AddNewReturnParams = true;
+            actExcel.SelectAllRows = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(string.Join(',', actExcel.ActReturnValues.Select(x => x.Actual).ToList()), "1,Mark,Cohen,2109 Fox Dr");
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:101]: this is not enough to validate only the "Actual" field, need to validate also the "Param" and "Path"

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadData;
+            actExcel.AddNewReturnParams = true;
+            actExcel.SelectAllRows = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            IExcelOperations excelOperations = new ExcelNPOIOperations();
+            DataTable dt = excelOperations.ReadData(excelPathWriteTemp, actExcel.SheetName, actExcel.SelectRowsWhere, actExcel.SelectAllRows);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:123]: this is Read Excel operation unit test, so why Assert validating with ReadExcel and not with Act.ReturnValues?

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SheetName), "Sheet1");
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SelectRowsWhere), "A2");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadCellData;
+            actExcel.AddNewReturnParams = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(actExcel.ActReturnValues.Count, 1);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:142]: why validating only number?

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+            actExcel.AddOrUpdateInputParamValueAndCalculatedValue(nameof(ActExcel.SheetName), "Sheet1");
+            actExcel.ExcelActionType = ActExcel.eExcelActionType.ReadCellData;
+            actExcel.AddNewReturnParams = true;
+            actExcel.SelectAllRows = true;
+
+            //Act
+            actExcel.Execute();
+
+            //Assert
+            Assert.AreEqual(actExcel.ActReturnValues.Count, 12);
```
Comment (MeniKadosh1) [inline Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:160]: why validating only number?

**Review by MeniKadosh1 — CHANGES_REQUESTED** [review]

File: Ginger/Ginger/Actions/ActionEditPages/ActExcelEditPage.xaml.cs (hunk 9 above)
Comment (uzanh) [reply Ginger/Ginger/Actions/ActionEditPages/ActExcelEditPage.xaml.cs:131]: yes

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 13 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:82]: done

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 14 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:101]: done

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 17 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs]: done

Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs]: done

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 12 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:62]: add param

File: Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs (hunk 16 above)
Comment (uzanh) [reply Ginger/GingerCoreNETUnitTest/LinuxTransformationTests/ActExcelTests.cs:146]: done

File: Ginger/GingerCoreCommon/InterfacesLib/IExcelOperations.cs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+        /// </summary>
+        /// <param name="filter"></param>
+        /// <param name="selectedRows"></param>
+        /// <param name="primaryKey"></param>
+        /// <param name="setData"></param>
+        /// <returns></returns>
+        DataTable ReadData(string fileName, string sheetName, string filter, bool selectedRows, string primaryKey, string setData);
+        bool WriteData(string fileName, string sheetName, string filter, string setDataUsed, List<Tuple<string, object>> updateCellValuesList, string primaryKey = null, string key = null);
+        DataTable ReadCellData(string fileName, string sheetName, string filter, bool selectedRows);
+        bool updateExcelData(string fileName, string sheetName, string filter, string setDataUsed, string primaryKey = null, string key = null);
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreCommon/InterfacesLib/IExcelOperations.cs:51]: done

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs:234]: done

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 19]
```diff
@@ -0,0 +1,562 @@
...
+                        for (int i = 0; i < excelDataTable.Columns.Count; i++)
+                        {
+                            AddOrUpdateReturnParamActualWithPath("Actual", ((object)r[i]).ToString(), (j + 1).ToString() + (i + 1).ToString());
+                        }
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                this.Status = Amdocs.Ginger.CoreNET.Execution.eRunStatus.Failed;
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs]: done

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 20]
```diff
@@ -0,0 +1,562 @@
...
+                Error = "No rows found in excel file matching criteria - ";
+            }
+
+        }
+
+        public List<string> GetSheets()
+        {
+            List<string> returnList = new List<string>();
+            var fileExtension = WorkSpace.Instance.SolutionRepository.ConvertSolutionRelativePath(ExcelFileName);
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs]: changed location and structure

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 21]
```diff
@@ -0,0 +1,562 @@
...
+
+                        }
+                        
+                    }
+                    bool isUpdated = true;
+                    if (SelectAllRows)
+                    {
+                        if (!String.IsNullOrWhiteSpace(SetDataUsed))
+                        {
+                            isUpdated = excelOperator.updateExcelData(GetExcelFileNameForDriver(), SheetName, SelectRowsWhere, SetDataUsed);
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs]: done

File: Ginger/GingerCoreNET/ActionsLib/ActExcel.cs [hunk 22]
```diff
@@ -0,0 +1,562 @@
...
+            }
+
+            if (excelDataTable.Rows.Count == 0)
+            {
+                this.Status = Amdocs.Ginger.CoreNET.Execution.eRunStatus.Failed;
+                Error = "No rows found in excel file matching criteria - ";
+            }
+        }
+
+        public bool check_file_open(String name)
```
[the remark this answers is not in the document]
Comment (uzanh) [reply Ginger/GingerCoreNET/ActionsLib/ActExcel.cs]: deleted

**Review by MeniKadosh1 — APPROVED** [review]

**Pull request merged**